src/librustc_mir/interpret/eval_context.rs

```diff
@@ -556,34 +570,40 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         for local in frame.locals {
             self.deallocate_local(local.value)?;
         }
-        // Validate the return value. Do this after deallocating so that we catch dangling
-        // references.
-        if let Some(return_place) = frame.return_place {
-            if M::enforce_validity(self) {
-                // Data got changed, better make sure it matches the type!
-                // It is still possible that the return place held invalid data while
-                // the function is running, but that's okay because nobody could have
-                // accessed that same data from the "outside" to observe any broken
-                // invariant -- that is, unless a function somehow has a ptr to
-                // its return place... but the way MIR is currently generated, the
-                // return place is always a local and then this cannot happen.
-                self.validate_operand(
-                    self.place_to_op(return_place)?,
-                    vec![],
-                    None,
-                    /*const_mode*/false,
-                )?;
+
+        // When we're popping a stack frame for unwinding purposes,
+        // we don't care at all about returning-related stuff (i.e. return_place
+        // and return_to_block), because we're not performing a return from this frame.
```

Comment:
Shouldn't we, like, jump to the "unwind" edge instead?

Reply:
That's handled in Miri [here](https://github.com/rust-lang/miri/pull/693/files#diff-6bac323bc5f550c3b0997b72a303e450R1105), by inspecting the current frame. I tried to have Miri handle as much of the unwinding logic as possible, to make it easier to modify in the future. Also, it seems unlikely that catching unwinds in `const` fns will ever be a useful thing to implement.

Reply:
The link does not work any more, but I assume you mean `unwind_stack`? I see now. So you rely on the fact that the suspended stack frames have their `block` and `stmt` set to the terminator they were executing, and you go on there.

TBH, I think it would be cleaner to add "unwind" terminators here. It would not be a big change I think, and it makes the entire thing much easier to understand than desperately trying to keep this out of `rustc_mir`. So unless @oli-obk objects, I think I'd prefer that.

Reply:
> unless @oli-obk objects

au contraire

If you want to keep that kind of information out of the engine, make `StackPopCleanup` generic over `Machine`, add a `StackPopCleanup` assoc type to `Machine` and add a `Machine` variant to `StackPopCleanup` that contains a single value of `Machine::StackPopCleanup` type. Then miri can fill in its own enum there.

You can even remove `pop_stack_frame_unwind` and make that a method on `Machine`.

Reply:
rustc now handles jumping to unwinding edges